### PR TITLE
Gorgoris/dao/invoice interface test

### PR DESCRIPTION
**Pull request by metagon**

### CODE REVIEW

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/Installment.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    String paymentInfo;
+    String type;
+    BigDecimal amount;
+
+    /**
+     * Creates a new installment object and assigns the given identity to it
+     * @param identity The identity of the installment
+     * @return New installment object with the provided identity
+     */
+    public static Installment withIdentity(Long identity) {
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/Installment.java]: long raw type

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/Invoice.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+import java.util.List;
+
+/**
+ * Domain object representing an invoice
+ */
+public class Invoice {
+    private Long identity;
+    private LocalDateTime time;
+    private BigDecimal gross;
+    private BigDecimal paid = BigDecimal.ZERO;
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/Invoice.java]: should be a boolean (with null as default value - we need null values for find)

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/Invoice.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    public BigDecimal getPaid() {
+        return paid;
+    }
+
+    /**
+     * Adds an amount that got paid to the total paid amount
+     * @param paid The amount that is going to be added to the
+     *             total paid amount
+     */
+    public void addPaid(BigDecimal paid) {
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/Invoice.java]: setPaid(Boolean paid)

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/InvoiceValidatorImpl.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
+package com.at.ac.tuwien.sepm.ss15.edulium.domain.validation.impl;
+
+import com.at.ac.tuwien.sepm.ss15.edulium.domain.Invoice;
+import com.at.ac.tuwien.sepm.ss15.edulium.domain.validation.ValidationException;
+import com.at.ac.tuwien.sepm.ss15.edulium.domain.validation.Validator;
+
+import java.math.BigDecimal;
+
+public class InvoiceValidatorImpl implements Validator<Invoice> {
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/InvoiceValidatorImpl.java]: remove public

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/TestInvoiceValidator.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+    @Autowired
+    Validator<Invoice> invoiceValidator;
+
+    @Test
+    public void testValidateForCreate_shouldAcceptInvoice() throws ValidationException {
+        Invoice invoice = new Invoice();
+        invoice.setTime(LocalDateTime.now());
+        invoice.setGross(new BigDecimal("23"));
+        invoice.setCreator(new User());
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/TestInvoiceValidator.java]: invalid user -> should throw (add a test case for it)

use User.withIdentity("testuser") instead

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/TestInvoiceValidator.java [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        Invoice invoice = null;
+
+        invoiceValidator.validateForCreate(invoice);
+    }
+
+    @Test(expected = ValidationException.class)
+    public void testValidateForCreate_shouldFailWithNullGrossAmount() throws ValidationException {
+        Invoice invoice = new Invoice();
+        invoice.setTime(LocalDateTime.now());
+
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/TestInvoiceValidator.java]: invoice.setCreator(User.withIdentity("testuser")); is missing

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/TestInvoiceValidator.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        invoiceValidator.validateForCreate(invoice);
+    }
+
+    @Test(expected = ValidationException.class)
+    public void testValidateForCreate_shouldFailWithNegativeGrossAmount() throws ValidationException {
+        Invoice invoice = new Invoice();
+        invoice.setTime(LocalDateTime.now());
+        invoice.setGross(new BigDecimal("-22"));
+
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/TestInvoiceValidator.java]: invoice.setCreator(User.withIdentity("testuser")); is missing

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/TestInvoiceValidator.java [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    @Test(expected = ValidationException.class)
+    public void testValidateForCreate_shouldFailWithPaidAmount() throws ValidationException {
+        Invoice invoice = new Invoice();
+        invoice.setTime(LocalDateTime.now());
+        invoice.setGross(new BigDecimal("15"));
+        invoice.addPaid(new BigDecimal("16"));
+
+        invoiceValidator.validateForCreate(invoice);
+    }
+
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/TestInvoiceValidator.java]: test case testValidatorForCreate_shouldFailWithNullTime is missing

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/TestInvoiceValidator.java [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+        invoiceValidator.validateForDelete(invoice);
+    }
+
+    @Test(expected = ValidationException.class)
+    public void testValidateForDelete_shouldFailWithNullIdentity() throws ValidationException {
+        Invoice invoice = new Invoice();
+
+        invoiceValidator.validateForDelete(invoice);
+    }
+}
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/TestInvoiceValidator.java]: identity test cases are missing (accept, null and null identity)

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestInvoiceDAO.java [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+    @Autowired
+    private DAO<Invoice> invoiceDAO;
+
+    @Test
+    public void testCreate_shouldAddObject() throws ValidationException, DAOException {
+        // GIVEN
+        Invoice invoice = new Invoice();
+        invoice.setTime(LocalDateTime.now());
+        invoice.setGross(new BigDecimal("15.6"));
+        invoice.setCreator(new User());
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestInvoiceDAO.java]: invalid user -> User.withIdentity

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestInvoiceDAO.java [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+        assertEquals(invoice, invoiceList.get(0));
+    }
+
+    @Test
+    public void testUpdate_shouldUpdateObject() throws ValidationException, DAOException {
+        // GIVEN
+        Invoice invoice = new Invoice();
+        invoice.setTime(LocalDateTime.now());
+        invoice.setGross(new BigDecimal("26.7"));
+        invoice.setCreator(new User());
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestInvoiceDAO.java]: invalid user -> User.withIdentity

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestInvoiceDAO.java [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+        assertEquals(invoice, invoiceList.get(0));
+    }
+
+    @Test
+    public void testDelete_shouldDeleteObject() throws ValidationException, DAOException {
+        // GIVEN
+        Invoice invoice = new Invoice();
+        invoice.setTime(LocalDateTime.now());
+        invoice.setGross(new BigDecimal("22.0"));
+        invoice.setCreator(new User());
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestInvoiceDAO.java]: invalid user -> User.withIdentity

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestInvoiceDAO.java [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+    }
+
+    @Test
+    public void testUpdate_shouldUpdateObject() throws ValidationException, DAOException {
+        // GIVEN
+        Invoice invoice = new Invoice();
+        invoice.setTime(LocalDateTime.now());
+        invoice.setGross(new BigDecimal("26.7"));
+        invoice.setCreator(new User());
+        invoiceDAO.create(invoice);
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestInvoiceDAO.java]: check if invoice is stored 

e.g. assertEquals(1, invoiceDAO.find(invoice).size());

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestInvoiceDAO.java [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+    }
+
+    @Test
+    public void testDelete_shouldDeleteObject() throws ValidationException, DAOException {
+        // GIVEN
+        Invoice invoice = new Invoice();
+        invoice.setTime(LocalDateTime.now());
+        invoice.setGross(new BigDecimal("22.0"));
+        invoice.setCreator(new User());
+        invoiceDAO.create(invoice);
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestInvoiceDAO.java]: check if invoice is stored

e.g. assertEquals(1, invoiceDAO.find(invoice).size());

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestInvoiceDAO.java [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+        inv3.setTime(now);
+        inv3.setGross(new BigDecimal("13.0"));
+        inv3.setCreator(new User());
+
+        invoiceDAO.create(inv1);
+        invoiceDAO.create(inv2);
+        invoiceDAO.create(inv3);
+
+        // WHEN
+        matcher.setIdentity(inv1.getIdentity());
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestInvoiceDAO.java]: remove the matcher and user Invoice.withIdentity instead

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestInvoiceDAO.java [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+        // WHEN
+        invoiceDAO.delete(invoice);
+
+        // THEN
+        Long identity = invoice.getIdentity();
+        assertTrue(invoiceDAO.find(Invoice.withIdentity(identity)).isEmpty());
+    }
+
+    @Test
+    public void testFind_shouldFindObject() throws ValidationException, DAOException {
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestInvoiceDAO.java]: shouldFindObjectsByIdentity

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestInvoiceDAO.java [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+
+        // THEN
+        assertEquals(3, all.size());
+        assertTrue(all.contains(inv1));
+        assertTrue(all.contains(inv2));
+        assertTrue(all.contains(inv3));
+    }
+
+    @Test
+    public void testGetAll_shouldReturnEmptyList() throws DAOException {
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestInvoiceDAO.java]: remove

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestInvoiceDAO.java [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+
+        // WHEN
+        List<Invoice> allInvoices = invoiceDAO.getAll();
+        assertNotNull(allInvoices);
+        assertTrue(allInvoices.isEmpty());
+
+        // THEN
+        List<Invoice> results = invoiceDAO.find(invoice);
+        assertTrue(results.isEmpty());
+    }
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestInvoiceDAO.java]: // GIVEN
 Invoice invoice = new Invoice();
 invoice.setIdentity(1l); // Arbitrary identity

     // WHEN
       List<Invoice> results = invoiceDAO.find(invoice);

   // THEN
    assertTrue(results.isEmpty());

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestInvoiceDAO.java [hunk 19]
```diff
@@ -0,0 +1,207 @@
...
+
+    @Test
+    public void testGetAll_shouldReturnEmptyList() throws DAOException {
+        // WHEN / THEN
+        List<Invoice> all = invoiceDAO.getAll();
+        assertNotNull(all);
+        assertTrue(all.isEmpty());
+    }
+}
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestInvoiceDAO.java]: history test cases are missing

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestInvoiceDAO.java [hunk 20]
```diff
@@ -0,0 +1,207 @@
...
+        invoiceDAO.create(invoice);
+
+        // WHEN
+        invoiceDAO.delete(invoice);
+
+        // THEN
+        Long identity = invoice.getIdentity();
+        assertTrue(invoiceDAO.find(Invoice.withIdentity(identity)).isEmpty());
+    }
+
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestInvoiceDAO.java]: missing test cases:
testDelete_deletingNullObjectShouldFail
testDelete_deletingNotPersistentObjectShouldFail

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestInvoiceDAO.java [hunk 21]
```diff
@@ -0,0 +1,207 @@
...
+        // WHEN
+        invoice.addPaid(new BigDecimal("28.0"));
+        invoiceDAO.update(invoice);
+
+        // THEN
+        List<Invoice> invoiceList = invoiceDAO.find(invoice);
+        assertEquals(1, invoiceList.size());
+        assertEquals(invoice, invoiceList.get(0));
+    }
+
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestInvoiceDAO.java]: missing test cases:
testUpdate_updatingNullObjectShouldFail
testUpdate_updatingNotPersistentObjectShouldFail
testUpdate_updatingObjectWithoutIdentityShouldFail

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestInvoiceDAO.java [hunk 22]
```diff
@@ -0,0 +1,207 @@
...
+        // THEN
+        // check if identity was set
+        Long identity = invoice.getIdentity();
+        assertNotNull(identity);
+
+        List<Invoice> invoiceList = invoiceDAO.find(Invoice.withIdentity(identity));
+        assertEquals(1, invoiceList.size());
+        assertEquals(invoice, invoiceList.get(0));
+    }
+
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/dao/TestInvoiceDAO.java]: missing test cases:
testCreate_addingNullObjectShouldFail
testCreate_addingInvalidObjectShouldFail

**Comment by metagon** [comment]
Going to add the missing test cases soon

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/Invoice.java [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+    /**
+     * Sets whether or not the invoice was paid
+     * @param paid Set to true if the invoice was paid and false otherwise
+     */
+    public void setPaid(Boolean paid) {
+        this.paid = paid;
+    }
+
+    /**
+     * @return Returns the creator of this invoice
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/Invoice.java]: (can be null)

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/TestInvoiceValidator.java [hunk 24]
```diff
@@ -0,0 +1,150 @@
...
+        Invoice invoice = new Invoice();
+        invoice.setTime(LocalDateTime.now());
+        invoice.setGross(new BigDecimal("-22"));
+        invoice.setCreator(User.withIdentity("TestUser"));
+
+        invoiceValidator.validateForCreate(invoice);
+    }
+
+    @Test(expected = ValidationException.class)
+    public void testValidateForCreate_shouldFailWithPaidSetToTrue() throws ValidationException {
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/TestInvoiceValidator.java]: paid will not be stored in the database, so ignore it for create/update

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/InvoiceValidatorImpl.java [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+     * @param invoice Object to be validated
+     * @throws ValidationException Thrown if the validation fails
+     */
+    @Override
+    public void validateForUpdate(Invoice invoice) throws ValidationException {
+        if (invoice == null) {
+            throw new ValidationException("Object must not be null");
+        }
+
+        if (invoice.getTime() == null || invoice.getGross() == null) {
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/InvoiceValidatorImpl.java]: split it up

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/InvoiceValidatorImpl.java [hunk 26]
```diff
@@ -0,0 +1,101 @@
...
+     * @param invoice Object to be validated
+     * @throws ValidationException Thrown if the validation fails
+     */
+    @Override
+    public void validateForCreate(Invoice invoice) throws ValidationException {
+        if (invoice == null) {
+            throw new ValidationException("Object must not be null");
+        }
+
+        if (invoice.getTime() == null || invoice.getGross() == null) {
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/InvoiceValidatorImpl.java]: split it up

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/InvoiceValidatorImpl.java [hunk 27]
```diff
@@ -0,0 +1,101 @@
...
+
+        if (invoice.getGross().compareTo(BigDecimal.ZERO) < 0) {
+            throw new ValidationException("The total gross amount cannot be negative");
+        }
+
+        if (invoice.getPaid() == null) {
+            throw new ValidationException("Paid must be either true of false");
+        }
+
+        if (invoice.getCreator() == null ||
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/InvoiceValidatorImpl.java]: use Validator<User> instead

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/InvoiceValidatorImpl.java [hunk 28]
```diff
@@ -0,0 +1,101 @@
...
+
+        if (invoice.getTime() == null || invoice.getGross() == null) {
+            throw  new ValidationException("Time and gross must not be null");
+        }
+
+        if (invoice.getGross().compareTo(BigDecimal.ZERO) < 0) {
+            throw new ValidationException("The total gross amount cannot be negative");
+        }
+
+        if (invoice.getPaid() == null) {
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/InvoiceValidatorImpl.java]: remove

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/InvoiceValidatorImpl.java [hunk 29]
```diff
@@ -0,0 +1,101 @@
...
+
+        if (invoice.getTime() == null || invoice.getGross() == null) {
+            throw  new ValidationException("Time and gross must not be null");
+        }
+
+        if (invoice.getGross().compareTo(BigDecimal.ZERO) < 0) {
+            throw new ValidationException("The total gross amount cannot be negative");
+        }
+
+        if (invoice.getPaid() != null && invoice.getPaid().equals(Boolean.TRUE)) {
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/InvoiceValidatorImpl.java]: remove

File: src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/InvoiceValidatorImpl.java [hunk 30]
```diff
@@ -0,0 +1,101 @@
...
+            throw new ValidationException("The invoice cannot be paid upon creation");
+        }
+
+        if (invoice.getCreator() == null ||
```
Comment (emmanuel099) [inline src/main/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/validation/impl/InvoiceValidatorImpl.java]: use Validator<User> instead

File: src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/TestInvoiceValidator.java [hunk 31]
```diff
@@ -0,0 +1,150 @@
...
+        invoiceValidator.validateForUpdate(invoice);
+    }
+
+    @Test(expected = ValidationException.class)
+    public void testValidateForUpdate_shouldFailWithNullIdentity() throws ValidationException {
+        Invoice invoice = new Invoice();
+
+        invoiceValidator.validateForUpdate(invoice);
+    }
+
```
Comment (emmanuel099) [inline src/test/java/com/at/ac/tuwien/sepm/ss15/edulium/domain/TestInvoiceValidator.java]: testValidateForUpdate test cases are missing:
shouldFailWithNullCreator
shouldFailWithNullTime
shouldFailWithNullGrossAmount
shouldFailWithNegativeGrossAmount

**Pull request merged**